fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/TreeLike.java

```diff
@@ -0,0 +1,266 @@
+     *        and an iterable of instances of type {@code F} that are the result of folding the children of that node.
+     *        The result of the folding function is also of type {@code F}.
+     * @param <M> the type parameter of the result of the mapping function
+     * @param <F> the type parameter of the result of the folding function
+     * @return an {@link Optional} containing the fold of the tree rooted at {@code this} if it exists,
+     *         {@code Optional.empty()} otherwise.
+     */
+    @Nullable
+    default <M, F> F foldNullable(@Nonnull final Function<T, M> mapFunction,
+                                  @Nonnull final BiFunction<M, Iterable<? extends F>, F> foldFunction) {
+        final M mappedThis = mapFunction.apply(getThis());
+        final List<F> foldedChildren = Lists.newArrayList();
+
+        for (final T child : getChildren()) {
+            foldedChildren.add(child.foldNullable(mapFunction, foldFunction));
+        }
+
+        return foldFunction.apply(mappedThis, foldedChildren);
+    }
+
+    /**
+     * Method that maps the tree-like rooted at {@code this} to another tree-like using a fold operation that folds into
+     * a {@code TreeLike<V>} if the fold exists.
+     * This method is normally used when an immutable tree-like structure needs to be <em>modified</em> in a
+     * persistent (copy-on-write) way meaning that parts or even all of the original tree rooted at {@code this}
+     * needs to be recreated using the {@link #withChildren(Iterable)} method.
+     * @param foldFunction a fold function that returns tree-like structures
+     * @param <V> type parameter constraint to {@code TreeLike<V>}
+     * @return an {@link Optional} of a new tree-like object that is result of the tree-map operation if the fold of the
+     *         tree rooted at {@code this} exists, {@code Optional.empty()} otherwise
+     */
+    @Nonnull
+    default <V extends TreeLike<V>> Optional<V> mapMaybe(@Nonnull final BiFunction<T, Iterable<? extends V>, V> foldFunction) {
+        return Optional.ofNullable(foldNullable(Function.identity(), foldFunction));
+    }
+
+    /**
+     * Method that maps the tree-like rooted at {@code this} to another tree-like of the same type. Unlike the other
+     * methods providing generic tree folds and maps, this method provides an easy way to mutate the leaves of the
+     * tree-like rooted at {@code this}.
+     * @param replaceOperator an operator that translates a tree-like of {@code T} into another tree-like of type
+     *        {@code T}.
+     * @return an {@link Optional} of a new tree-like object that is result of the tree-map operation if the fold of the
+     *         tree rooted at {@code this} exists, {@code Optional.empty()} otherwise
+     */
+    @SuppressWarnings("PMD.CompareObjectsWithEquals")
+    @Nonnull
+    default Optional<T> mapLeavesMaybe(@Nonnull final UnaryOperator<T> replaceOperator) {
```

Comment:
This method name seems a little vague to me. Is there some precedent for something that better captures what it does? Like maybe from Lens or Monocle or some other similar system?

Reply:
Hmm there is nothing that fits --  maybe it's like a prism but not sure. I renamed it to `replaceLeavesMaybe` which seems better.

Reply:
✅